modules/tensor_mechanics/src/kernels/TotalLagrangianStressDivergence.C

```diff
@@ -0,0 +1,132 @@
+#include "TotalLagrangianStressDivergence.h"
+
+registerMooseObject("TensorMechanicsApp", TotalLagrangianStressDivergence);
+
+InputParameters
+TotalLagrangianStressDivergence::validParams()
+{
+  InputParameters params = LagrangianStressDivergenceBase::validParams();
+
+  params.addClassDescription("Total Lagrangian stress equilibrium kernel");
+
+  // This kernel requires use_displaced_mesh to be off
+  params.suppressParameter<bool>("use_displaced_mesh");
+
+  return params;
+}
+
+TotalLagrangianStressDivergence::TotalLagrangianStressDivergence(const InputParameters & parameters)
+  : LagrangianStressDivergenceBase(parameters),
+    _pk1(getMaterialPropertyByName<RankTwoTensor>(_base_name + "pk1_stress")),
+    _dpk1(getMaterialPropertyByName<RankFourTensor>(_base_name + "pk1_jacobian")),
+    _avg_grad_trial(_phi.size()),
+    _uF(getMaterialPropertyByName<RankTwoTensor>(_base_name + "unstabilized_deformation_gradient")),
+    _aF(getMaterialPropertyByName<RankTwoTensor>(_base_name + "avg_deformation_gradient"))
+{
+}
+
+Real
+TotalLagrangianStressDivergence::computeQpResidual()
+{
+  return _pk1[_qp].doubleContraction(testGrad(_component));
+}
+
+Real
+TotalLagrangianStressDivergence::computeQpJacobian()
+{
+  return materialJacobian(testGrad(_component), trialGrad(_component));
+}
+
+Real
+TotalLagrangianStressDivergence::computeQpOffDiagJacobian(unsigned int jvar)
+{
+  for (unsigned int cc = 0; cc < _ndisp; cc++)
+    if (jvar == _disp_nums[cc])
+      return materialJacobian(testGrad(_component), trialGrad(cc));
+
```

Comment:
I see that in `ComputeLagrangianStrain` you included eigenstrains, so I would also expect to see derivatives of eigenstrains w.r.t. other variables here, e.g. `deigenstrain_dtemperature`.

Reply:
Yes, I should do this.

Reply:
Done.